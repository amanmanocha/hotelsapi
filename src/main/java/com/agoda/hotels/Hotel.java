package com.agoda.hotels;

import javax.persistence.Column;
import javax.persistence.Entity;
import javax.persistence.Id;

@Entity(name = "hotels")
public class Hotel {
  @Id
  @Column(name = "hotelid")
  private String hotelId;
  @Column(name = "city")
  private String city;
  @Column(name = "room")
  private String room;
  @Column(name = "price")
  private int price;

  Hotel() {
  }

  public Hotel(String hotelId, String city, String room, int price) {
    super();
    this.hotelId = hotelId;
    this.city = city;
    this.room = room;
    this.price = price;
  }

  public String getHotelId() {
    return hotelId;
  }

  public String getCity() {
    return city;
  }

  public String getRoom() {
    return room;
  }

  public int getPrice() {
    return price;
  }

}
