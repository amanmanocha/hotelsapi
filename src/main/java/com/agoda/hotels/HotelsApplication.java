package com.agoda.hotels;

import org.springframework.beans.factory.annotation.Autowired;
import org.springframework.boot.SpringApplication;
import org.springframework.boot.autoconfigure.SpringBootApplication;
import org.springframework.context.annotation.Bean;
import org.springframework.web.servlet.HandlerInterceptor;
import org.springframework.web.servlet.handler.MappedInterceptor;

@SpringBootApplication
public class HotelsApplication {
  private final HandlerInterceptor rateLimitInterceptor;
  
  @Autowired
  public HotelsApplication(HandlerInterceptor interceptor) {
    this.rateLimitInterceptor = interceptor;
  }

  public static void main(String[] args) {
    SpringApplication.run(HotelsApplication.class, args);
  }
  
  @Bean
  public MappedInterceptor interceptors() {
    return new MappedInterceptor(new String[] { "/**" }, rateLimitInterceptor);
  }
}