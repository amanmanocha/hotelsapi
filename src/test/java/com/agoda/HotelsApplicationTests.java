package com.agoda;

import static org.approvaltests.Approvals.verify;
import static org.springframework.test.web.servlet.request.MockMvcRequestBuilders.get;
import static org.springframework.test.web.servlet.result.MockMvcResultMatchers.content;
import static org.springframework.test.web.servlet.result.MockMvcResultMatchers.status;

import java.nio.charset.Charset;

import org.junit.Before;
import org.junit.Test;
import org.junit.runner.RunWith;
import org.springframework.beans.factory.annotation.Autowired;
import org.springframework.boot.test.context.SpringBootTest;
import org.springframework.http.MediaType;
import org.springframework.test.context.junit4.SpringRunner;
import org.springframework.test.context.web.WebAppConfiguration;
import org.springframework.test.web.servlet.MockMvc;
import org.springframework.test.web.servlet.MvcResult;
import org.springframework.test.web.servlet.setup.MockMvcBuilders;
import org.springframework.web.context.WebApplicationContext;

import com.agoda.hotels.HotelsApplication;
@RunWith(SpringRunner.class)
@SpringBootTest(classes=HotelsApplication.class)
@WebAppConfiguration
public class HotelsApplicationTests {
  private static final String TEST_API_KEY = "1";

  private final MediaType EXPECTED_CONTENT_TYPE = new MediaType("application", "hal+json", Charset.forName("UTF-8"));

  @Autowired
  private WebApplicationContext webApplicationContext;

  private MockMvc mockMvc;

  @Before
  public void setup() {
      mockMvc = MockMvcBuilders.webAppContextSetup(webApplicationContext)
              .build();
  }

  @Test public void 
  hotels_get_end_point_should_return_all_hotels() throws Exception {
      MvcResult result = mockMvc.perform(get("/hotels").header("x-api-key", TEST_API_KEY))
                                .andExpect(status().isOk())
                                .andExpect(content().contentType(EXPECTED_CONTENT_TYPE))
                                .andReturn();
      verify(result.getResponse().getContentAsString());
  }
  
  @Test public void 
  hotels_get_with_sort_by_price_should_return_all_hotels_sorted_by_price() throws Exception {
      MvcResult result = mockMvc.perform(get("/hotels?sort=price").header("x-api-key", TEST_API_KEY))
                                .andExpect(status().isOk())
                                .andExpect(content().contentType(EXPECTED_CONTENT_TYPE))
                                .andReturn();
      verify(result.getResponse().getContentAsString());
  }
  
  @Test public void 
  hotels_get_with_sort_by_price_desc_should_return_all_hotels_sorted_by_price_desc() throws Exception {
      MvcResult result = mockMvc.perform(get("/hotels?sort=price,desc").header("x-api-key", TEST_API_KEY))
                                .andExpect(status().isOk())
                                .andExpect(content().contentType(EXPECTED_CONTENT_TYPE))
                                .andReturn();
      verify(result.getResponse().getContentAsString());
  }
  
  @Test public void 
  hotels_get_with_search_by_city_should_return_hotels_for_city_only() throws Exception {
      MvcResult result = mockMvc.perform(get("/hotels/search/findByCity?city=Bangkok").header("x-api-key", TEST_API_KEY))
                                .andExpect(status().isOk())
                                .andExpect(content().contentType(EXPECTED_CONTENT_TYPE))
                                .andReturn();
      verify(result.getResponse().getContentAsString());
  }
  
  @Test public void 
  hotels_get_without_api_key_returns_forbidden() throws Exception {
      mockMvc.perform(get("/hotels"))
             .andExpect(status().isForbidden());
  }
}
