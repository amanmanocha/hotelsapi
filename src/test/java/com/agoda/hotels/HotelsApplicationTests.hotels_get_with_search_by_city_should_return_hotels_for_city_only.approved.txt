{
  "_embedded" : {
    "hotels" : [ {
      "city" : "Bangkok",
      "room" : "Deluxe",
      "price" : 1000,
      "_links" : {
        "self" : {
          "href" : "http://localhost/hotels/1"
        },
        "hotel" : {
          "href" : "http://localhost/hotels/1"
        }
      }
    }, {
      "city" : "Bangkok",
      "room" : "Superior",
      "price" : 2000,
      "_links" : {
        "self" : {
          "href" : "http://localhost/hotels/6"
        },
        "hotel" : {
          "href" : "http://localhost/hotels/6"
        }
      }
    }, {
      "city" : "Bangkok",
      "room" : "Superior",
      "price" : 2400,
      "_links" : {
        "self" : {
          "href" : "http://localhost/hotels/8"
        },
        "hotel" : {
          "href" : "http://localhost/hotels/8"
        }
      }
    }, {
      "city" : "Bangkok",
      "room" : "Deluxe",
      "price" : 60,
      "_links" : {
        "self" : {
          "href" : "http://localhost/hotels/11"
        },
        "hotel" : {
          "href" : "http://localhost/hotels/11"
        }
      }
    }, {
      "city" : "Bangkok",
      "room" : "Sweet Suite",
      "price" : 25000,
      "_links" : {
        "self" : {
          "href" : "http://localhost/hotels/14"
        },
        "hotel" : {
          "href" : "http://localhost/hotels/14"
        }
      }
    }, {
      "city" : "Bangkok",
      "room" : "Deluxe",
      "price" : 900,
      "_links" : {
        "self" : {
          "href" : "http://localhost/hotels/15"
        },
        "hotel" : {
          "href" : "http://localhost/hotels/15"
        }
      }
    }, {
      "city" : "Bangkok",
      "room" : "Sweet Suite",
      "price" : 5300,
      "_links" : {
        "self" : {
          "href" : "http://localhost/hotels/18"
        },
        "hotel" : {
          "href" : "http://localhost/hotels/18"
        }
      }
    } ]
  },
  "_links" : {
    "self" : {
      "href" : "http://localhost/hotels/search/findByCity?city=Bangkok"
    }
  }
}